src/compiler/crystal/macros.cr

```diff
@@ -153,6 +153,35 @@ module Crystal::Macros
   def host_flag?(name) : BoolLiteral
   end
 
+  # Parses *type_name* into a `Path` or `Generic` (also used for unions).
+  # The `Path#resolve` or `Generic#resolve` method could then be used
+  # to resolve the value into a `TypeNode`, if the *type_name* represents a type,
+  # otherwise the value of the constant.
+  #
+  # A compile time error is raised if the type/constant does not actually exist,
+  # or if a required generic argument was not provided.
+  #
+  # ```
+  # class Foo; end
+  #
+  # struct Some::Namespace::Foo; end
+  #
+  # module Bar(T); end
+  #
+  # MY_CONST = 1234
+  #
+  # {{ parse_type("Foo").resolve.class? }}                                   # => true
+  # {{ parse_type("Some::Namespace::Foo").resolve.struct? }}                 # => true
+  # {{ parse_type("Foo|Some::Namespace::Foo").resolve.union_types.size }}    # => 2
+  # {{ parse_type("Bar(Int32)|Foo").resolve.union_types[0].type_vars.size }} # => 1
+  # {{ parse_type("MY_CONST").resolve }}                                     # => 1234
+  #
+  # {{ parse_type("MissingType").resolve }}   # => Error: undefined constant MissingType
+  # {{ parse_type("UNKNOWN_CONST").resolve }} # => Error: undefined constant UNKNOWN_CONST
+  # ```
+  def parse_type(type_name : StringLiteral) : Path | Generic
```

Comment:
Given this comment:

https://github.com/crystal-lang/crystal/pull/11126#issuecomment-953867122

it might be better to say that this returns `ASTNode`.

Reply:
Hmm, thoughts on just adding them to the union to be more clear what it could return?

Reply:
This is just a minor thing, so we can merge this right away and always improve the docs later on.

Reply:
I already had it done, plus addressed your other comment, so I just pushed it up. Should be good to go now.